targets/TARGET_RDA/TARGET_UNO_91H/trng_api.c

```diff
@@ -0,0 +1,97 @@
+/*
+ *  Hardware entropy collector for the UNO_91H, using RDA's TRNG
+ *
+ *  Copyright (C) 2006-2016, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#if defined(DEVICE_TRNG)
+
+#include "cmsis.h"
+#include "trng_api.h"
+
+#define rRNG_CLKGATE        (RDA_SCU->CLKGATE0)
+#define rRNG_CLKGEN         (RDA_DMACFG->dma_func_ctrl)
+#define rRNG_INTSTATE       (RDA_DMACFG->dma_int_out)
+#define rTRNG_CTRL          (RDA_RNG->TCTRL)
+#define rTRNG_DATA0         (RDA_RNG->TD0)
+#define rTRNG_DATA1         (RDA_RNG->TD1)
+#define rPRNG_DATA          (RDA_RNG->PD)
+
+/** trng_get_byte
+ *  @brief Get one byte of entropy from the RNG, assuming it is up and running.
+ *  @param obj TRNG obj
+ *  @param pointer to the hardware generated random byte.
+ */
+static void trng_get_byte(trng_t *obj, unsigned char *byte )
+{
+    if(0U == obj->byte_idx) {
+        int n;
+        /* Delay for PRNG data ready */
+        for(n = 0; n < 4096; n++);
```

Comment:
This might be optimized away by the compiler, and even if it's not, it's not a reliable way to enforce a delay. Please clarify and quantify the need for a delay here and use an appropriate waiting mechanism.

Reply:
The data rate generated by TRNG analog module is 1M(bps), so the TRNG_CTRL outputs 32-bit data in 1/32M(bps). There is no complete-flag or Interrupt mechanism, the delay should be implemented by software. How about using us_ticker_read()? @hanno-arm

Reply:
@RdaZhongyao It is surprising to hear that the TRNG does neither support notification via interrupt nor via complete flag. If the only information available is the output rate, a static and reliable wait together with a check that the TRNG is in ordinary operation (if there's a register for that) seems to be the best option that remains. @mazimkhan @0xc0170 Ping.

Reply:
us_ticker_read() would be better in this case than just a sw loop